src/libraries/System.Linq/src/System/Linq/Iterator.SpeedOpt.cs

```diff
@@ -0,0 +1,71 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+
+namespace System.Linq
+{
+    public static partial class Enumerable
+    {
+        internal abstract partial class Iterator<TSource>
+        {
+            /// <summary>
+            /// Produce an array of the sequence through an optimized path.
+            /// </summary>
+            /// <returns>The array.</returns>
+            public abstract TSource[] ToArray();
+
+            /// <summary>
+            /// Produce a <see cref="List{TSource}"/> of the sequence through an optimized path.
+            /// </summary>
+            /// <returns>The <see cref="List{TSource}"/>.</returns>
+            public abstract List<TSource> ToList();
+
+            /// <summary>
+            /// Returns the count of elements in the sequence.
+            /// </summary>
+            /// <param name="onlyIfCheap">If true then the count should only be calculated if doing
+            /// so is quick (sure or likely to be constant time), otherwise -1 should be returned.</param>
+            /// <returns>The number of elements.</returns>
+            public abstract int GetCount(bool onlyIfCheap);
+
+            /// <summary>
+            /// Creates a new iterator that skips the specified number of elements from this sequence.
+            /// </summary>
+            /// <param name="count">The number of elements to skip.</param>
+            /// <returns>An <see cref="Iterator{TSource}"/> with the first <paramref name="count"/> items removed, or null if known empty.</returns>
+            public virtual Iterator<TSource>? Skip(int count) => new IEnumerableSkipTakeIterator<TSource>(this, count, -1);
```

Comment:
A few years back the `Skip` and `Take` implementations were consolidated behind a single `Take(Range)` API. Would it make sense to have the Partition implementation reflect this as well?

Reply:
We could look at that as a subsequent change. I don't know how commonly used that method is, though, and at least currently it's not a choke point, so getting everything to go through it would require some refactoring. There might also be some extra costs incurred, in particular in support of handling from-end indices.